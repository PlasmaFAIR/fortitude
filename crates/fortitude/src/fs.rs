use std::fmt::{Display, Formatter};
use std::hash::Hasher;
use std::ops::Deref;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use globset::{Glob, GlobSet, GlobSetBuilder};
use ignore::{types::TypesBuilder, WalkBuilder};
use itertools::Itertools;
use log::debug;
use path_absolutize::Absolutize;
use ruff_cache::{CacheKey, CacheKeyHasher};
use ruff_macros::CacheKey;
use serde::{de, Deserialize, Deserializer, Serialize};

use crate::registry::Rule;
use crate::rule_selector::CompiledPerFileIgnoreList;
use crate::settings::FileResolverSettings;

#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord, Serialize, CacheKey)]
pub enum FilePattern {
    Builtin(&'static str),
    User(String, PathBuf),
}

impl FilePattern {
    const EXPECTED_PATTERN: &'static str = "<FilePattern>";

    pub fn add_to(self, builder: &mut GlobSetBuilder) -> anyhow::Result<()> {
        match self {
            FilePattern::Builtin(pattern) => {
                builder.add(Glob::from_str(pattern)?);
            }
            FilePattern::User(pattern, absolute) => {
                // Add the absolute path.
                builder.add(Glob::new(&absolute.to_string_lossy())?);

                // Add basename path.
                if !pattern.contains(std::path::MAIN_SEPARATOR) {
                    builder.add(Glob::new(&pattern)?);
                }
            }
        }
        Ok(())
    }
}

impl FromStr for FilePattern {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let pattern = s.to_string();
        let absolute = normalize_path(&pattern);
        Ok(Self::User(pattern, absolute))
    }
}

impl<'de> Deserialize<'de> for FilePattern {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let str_result = String::deserialize(deserializer)?;
        Self::from_str(str_result.as_str()).map_err(|_| {
            de::Error::invalid_value(
                de::Unexpected::Str(str_result.as_str()),
                &Self::EXPECTED_PATTERN,
            )
        })
    }
}

impl Display for FilePattern {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{:?}",
            match self {
                Self::Builtin(pattern) => pattern,
                Self::User(pattern, _) => pattern.as_str(),
            }
        )
    }
}

#[derive(Debug, Clone, Default)]
pub struct FilePatternSet {
    set: GlobSet,
    cache_key: u64,
    // This field is only for displaying the internals
    // of `set`.
    #[allow(clippy::used_underscore_binding)]
    _set_internals: Vec<FilePattern>,
}

impl Display for FilePatternSet {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        if self._set_internals.is_empty() {
            write!(f, "[]")?;
        } else {
            writeln!(f, "[")?;
            for pattern in &self._set_internals {
                writeln!(f, "\t{pattern},")?;
            }
            write!(f, "]")?;
        }
        Ok(())
    }
}

impl Deref for FilePatternSet {
    type Target = GlobSet;

    fn deref(&self) -> &Self::Target {
        &self.set
    }
}

impl CacheKey for FilePatternSet {
    fn cache_key(&self, state: &mut CacheKeyHasher) {
        state.write_usize(self.set.len());
        state.write_u64(self.cache_key);
    }
}

impl FilePatternSet {
    pub fn try_from_iter<I>(patterns: I) -> Result<Self, anyhow::Error>
    where
        I: IntoIterator<Item = FilePattern>,
    {
        let mut builder = GlobSetBuilder::new();
        let mut hasher = CacheKeyHasher::new();

        let mut _set_internals = vec![];

        for pattern in patterns {
            _set_internals.push(pattern.clone());
            pattern.cache_key(&mut hasher);
            pattern.add_to(&mut builder)?;
        }
        let set = builder.build()?;
        Ok(FilePatternSet {
            set,
            cache_key: hasher.finish(),
            _set_internals,
        })
    }

    pub fn matches<P: AsRef<Path>>(&self, path: P) -> bool {
        match std::path::absolute(path.as_ref()) {
            Ok(path) => match path.clone().file_name() {
                Some(basename) => self.set.is_match(path) || self.set.is_match(basename),
                None => false,
            },
            _ => false,
        }
    }

    pub fn ancestor_matches<P: AsRef<Path>, R: AsRef<Path>>(
        &self,
        path: P,
        project_root: R,
    ) -> bool {
        let project_root = project_root.as_ref();
        match std::path::absolute(path.as_ref()) {
            Ok(path) => path
                .ancestors()
                .take_while(|ancestor| *ancestor != project_root)
                .any(|ancestor| match ancestor.file_name() {
                    Some(basename) => self.set.is_match(ancestor) || self.set.is_match(basename),
                    None => false,
                }),
            _ => false,
        }
    }
}

/// Create a set with codes matching the pattern/code pairs.
pub(crate) fn ignores_from_path(path: &Path, ignore_list: &CompiledPerFileIgnoreList) -> Vec<Rule> {
    let file_name = path.file_name().expect("Unable to parse filename");
    ignore_list
        .iter()
        .filter_map(|entry| {
            if entry.basename_matcher.is_match(file_name) || entry.absolute_matcher.is_match(path) {
                if entry.negated {
                    None
                } else {
                    Some(&entry.rules)
                }
            } else if entry.negated {
                Some(&entry.rules)
            } else {
                None
            }
        })
        .flatten()
        .collect()
}

/// Convert any path to an absolute path (based on the current working
/// directory).
pub fn normalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path = path.as_ref();
    if let Ok(path) = path.absolutize() {
        return path.to_path_buf();
    }
    path.to_path_buf()
}

/// Convert any path to an absolute path (based on the specified project root).
pub fn normalize_path_to<P: AsRef<Path>, R: AsRef<Path>>(path: P, project_root: R) -> PathBuf {
    let path = path.as_ref();
    if let Ok(path) = path.absolutize_from(project_root.as_ref()) {
        return path.to_path_buf();
    }
    path.to_path_buf()
}

/// Convert an absolute path to be relative to the current working directory.
pub fn relativize_path<P: AsRef<Path>>(path: P) -> String {
    let path = path.as_ref();

    #[cfg(target_arch = "wasm32")]
    let cwd = Path::new(".");
    #[cfg(not(target_arch = "wasm32"))]
    let cwd = path_absolutize::path_dedot::CWD.as_path();

    if let Ok(path) = path.strip_prefix(cwd) {
        return format!("{}", path.display());
    }
    format!("{}", path.display())
}

/// Convert an absolute path to be relative to the specified project root.
#[allow(dead_code)]
pub fn relativize_path_to<P: AsRef<Path>, R: AsRef<Path>>(path: P, project_root: R) -> String {
    format!(
        "{}",
        pathdiff::diff_paths(&path, project_root)
            .expect("Could not diff paths")
            .display()
    )
}

/// Default extensions to check
pub const FORTRAN_EXTS: &[&str] = &[
    "f90", "F90", "f95", "F95", "f03", "F03", "f08", "F08", "f18", "F18", "f23", "F23",
];

// Default paths to exclude when searching paths
pub(crate) static EXCLUDE_BUILTINS: &[FilePattern] = &[
    FilePattern::Builtin(".git"),
    FilePattern::Builtin(".git-rewrite"),
    FilePattern::Builtin(".hg"),
    FilePattern::Builtin(".svn"),
    FilePattern::Builtin("venv"),
    FilePattern::Builtin(".venv"),
    FilePattern::Builtin("pyenv"),
    FilePattern::Builtin(".pyenv"),
    FilePattern::Builtin(".eggs"),
    FilePattern::Builtin("site-packages"),
    FilePattern::Builtin(".vscode"),
    FilePattern::Builtin("build"),
    FilePattern::Builtin("_build"),
    FilePattern::Builtin("dist"),
    FilePattern::Builtin("_dist"),
];

/// Returns the default set of files if none are provided, otherwise
/// returns a list with just the current working directory.
fn resolve_default_files(files: &[PathBuf], is_stdin: bool) -> Vec<PathBuf> {
    if files.is_empty() {
        if is_stdin {
            vec![Path::new("-").to_path_buf()]
        } else {
            vec![Path::new(".").to_path_buf()]
        }
    } else {
        files.to_vec()
    }
}

/// Expand the input list of files to include all Fortran files.
pub fn get_files(resolver: &FileResolverSettings, is_stdin: bool) -> anyhow::Result<Vec<PathBuf>> {
    debug!("Gathering files");
    debug!("Project root: {:?}", resolver.project_root);
    let paths = resolve_default_files(&resolver.files, is_stdin);

    // Normalise all paths and remove duplicates.
    // If exclude_mode is set to Force, remove paths that match the exclude patterns.
    let paths: Vec<_> = if resolver.force_exclude {
        let (excluded, paths): (Vec<_>, Vec<_>) =
            paths.iter().map(normalize_path).unique().partition(|p| {
                resolver
                    .excludes
                    .ancestor_matches(p, &resolver.project_root)
            });
        if !excluded.is_empty() {
            debug!("Force excluded paths: {:?}", excluded);
        }
        paths
    } else {
        paths.iter().map(normalize_path).unique().collect()
    };
    debug!("Paths provided: {:?}", paths);

    // The remaining non-directory paths are always included; split into directories and files.
    // Note that this includes paths that do not exist, as these should be reported to the user.
    let (dirs, files): (Vec<_>, Vec<_>) = paths.into_iter().partition(|p| p.is_dir());

    let excludes = resolver.excludes.clone();

    // Collect all files from directories
    let dir_contents = if let Some((first_dir, rest)) = dirs.split_first() {
        // Create a directory walker that follows exclude patterns
        let mut builder = WalkBuilder::new(first_dir);
        for path in rest {
            builder.add(path);
        }
        builder.standard_filters(resolver.respect_gitignore);
        builder.hidden(false);
        builder.filter_entry(move |e| !excludes.matches(e.path()));

        // Add file type filter for provided file extensions
        // Directories will be skipped
        let mut file_types = TypesBuilder::new();
        for ext in &resolver.file_extensions {
            file_types.add(ext.as_ref(), format!("*.{}", ext).as_str())?;
        }
        file_types.select("all");
        builder.types(file_types.build()?);

        // Collect all valid files from directories
        builder
            .build()
            .filter_map(|p| p.ok()) // skip dirs if user doesn't have permission
            .map(|p| p.into_path())
            .filter(|p| !p.is_dir())
            .collect()
    } else {
        // No dirs remain after removing excludes and splitting into dirs and files
        vec![]
    };

    // Return all files found
    Ok(files.into_iter().chain(dir_contents).collect())
}
