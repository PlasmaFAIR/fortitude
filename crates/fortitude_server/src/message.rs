use anyhow::Context;
use lsp_types::notification::Notification;
use std::sync::OnceLock;

use crate::server::ClientSender;

static MESSENGER: OnceLock<ClientSender> = OnceLock::new();

pub(crate) fn init_messenger(client_sender: ClientSender) {
    MESSENGER
        .set(client_sender)
        .expect("messenger should only be initialized once");
}

pub(super) fn try_show_message(
    message: String,
    message_type: lsp_types::MessageType,
) -> crate::Result<()> {
    MESSENGER
        .get()
        .ok_or_else(|| anyhow::anyhow!("messenger not initialized"))?
        .send(lsp_server::Message::Notification(
            lsp_server::Notification {
                method: lsp_types::notification::ShowMessage::METHOD.into(),
                params: serde_json::to_value(lsp_types::ShowMessageParams {
                    typ: message_type,
                    message,
                })?,
            },
        ))
        .context("Failed to send message")?;

    Ok(())
}
