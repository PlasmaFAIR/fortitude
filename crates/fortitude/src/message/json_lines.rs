// Adapted from from ruff
// Copyright 2022 Charles Marsh
// SPDX-License-Identifier: MIT

use std::io::Write;

use crate::message::json::message_to_json_value;
use crate::message::Emitter;

use super::DiagnosticMessage;

#[derive(Default)]
pub struct JsonLinesEmitter;

impl Emitter for JsonLinesEmitter {
    fn emit(
        &mut self,
        writer: &mut dyn Write,
        messages: &[DiagnosticMessage],
    ) -> anyhow::Result<()> {
        for message in messages {
            serde_json::to_writer(&mut *writer, &message_to_json_value(message))?;
            writer.write_all(b"\n")?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::json_lines::JsonLinesEmitter;
    use crate::message::tests::{capture_emitter_output, create_messages};

    #[test]
    fn output() {
        let mut emitter = JsonLinesEmitter;
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
