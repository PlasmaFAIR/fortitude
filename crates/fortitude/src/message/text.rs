// Adapted from from ruff
// Copyright 2022 Charles Marsh
// SPDX-License-Identifier: MIT

use std::fmt::{Display, Formatter};
use std::io::Write;

use annotate_snippets::{Level, Renderer, Snippet};
use bitflags::bitflags;
use colored::Colorize;

use ruff_source_file::OneIndexed;
use ruff_text_size::{Ranged, TextRange};

use crate::fs::relativize_path;
// use crate::line_width::{IndentWidth, LineWidthBuilder};
use crate::message::diff::Diff;
use crate::message::Emitter;
use crate::settings::UnsafeFixes;
use crate::text_helpers::ShowNonprinting;

use super::DiagnosticMessage;

bitflags! {
    #[derive(Default)]
    struct EmitterFlags: u8 {
        /// Whether to show the fix status of a diagnostic.
        const SHOW_FIX_STATUS    = 0b0000_0001;
        /// Whether to show the diff of a fix, for diagnostics that have a fix.
        const SHOW_FIX_DIFF      = 0b0000_0010;
        /// Whether to show the source code of a diagnostic.
        const SHOW_SOURCE        = 0b0000_0100;
    }
}

#[derive(Default)]
pub struct TextEmitter {
    flags: EmitterFlags,
    unsafe_fixes: UnsafeFixes,
}

impl TextEmitter {
    #[must_use]
    pub fn with_show_fix_status(mut self, show_fix_status: bool) -> Self {
        self.flags
            .set(EmitterFlags::SHOW_FIX_STATUS, show_fix_status);
        self
    }

    #[must_use]
    pub fn with_show_fix_diff(mut self, show_fix_diff: bool) -> Self {
        self.flags.set(EmitterFlags::SHOW_FIX_DIFF, show_fix_diff);
        self
    }

    #[must_use]
    pub fn with_show_source(mut self, show_source: bool) -> Self {
        self.flags.set(EmitterFlags::SHOW_SOURCE, show_source);
        self
    }

    #[must_use]
    pub fn with_unsafe_fixes(mut self, unsafe_fixes: UnsafeFixes) -> Self {
        self.unsafe_fixes = unsafe_fixes;
        self
    }
}

impl Emitter for TextEmitter {
    fn emit(
        &mut self,
        writer: &mut dyn Write,
        messages: &[DiagnosticMessage],
    ) -> anyhow::Result<()> {
        for message in messages {
            write!(
                writer,
                "{path}{sep}",
                path = relativize_path(message.filename()).bold(),
                sep = ":".cyan(),
            )?;

            let start_location = message.compute_start_location();

            let diagnostic_location = start_location;

            writeln!(
                writer,
                "{row}{sep}{col}{sep} {code_and_body}",
                row = diagnostic_location.row,
                col = diagnostic_location.column,
                sep = ":".cyan(),
                code_and_body = RuleCodeAndBody {
                    message,
                    show_fix_status: self.flags.intersects(EmitterFlags::SHOW_FIX_STATUS),
                    unsafe_fixes: self.unsafe_fixes,
                }
            )?;

            if self.flags.intersects(EmitterFlags::SHOW_SOURCE) {
                // The `0..0` range is used to highlight file-level diagnostics.
                if message.range() != TextRange::default() {
                    writeln!(writer, "{}", MessageCodeFrame { message })?;
                }
            }

            if self.flags.intersects(EmitterFlags::SHOW_FIX_DIFF) {
                if let Some(diff) = Diff::from_message(message) {
                    writeln!(writer, "{diff}")?;
                }
            }
        }

        Ok(())
    }
}

pub(super) struct RuleCodeAndBody<'a> {
    pub(crate) message: &'a DiagnosticMessage,
    pub(crate) show_fix_status: bool,
    pub(crate) unsafe_fixes: UnsafeFixes,
}

impl Display for RuleCodeAndBody<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        if self.show_fix_status {
            if let Some(fix) = self.message.fix() {
                // Do not display an indicator for inapplicable fixes
                if fix.applies(self.unsafe_fixes.required_applicability()) {
                    if let Some(rule) = self.message.rule() {
                        write!(f, "{} ", rule.noqa_code().to_string().red().bold())?;
                    }
                    return write!(
                        f,
                        "{fix}{body}",
                        fix = format_args!("[{}] ", "*".cyan()),
                        body = self.message.body(),
                    );
                }
            }
        };

        if let Some(rule) = self.message.rule() {
            write!(
                f,
                "{code} {body}",
                code = rule.noqa_code().to_string().red().bold(),
                body = self.message.body(),
            )
        } else {
            f.write_str(self.message.body())
        }
    }
}

pub(super) struct MessageCodeFrame<'a> {
    pub(crate) message: &'a DiagnosticMessage,
}

impl Display for MessageCodeFrame<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let source_code = self.message.source_file().to_source_code();

        let content_start_index = source_code.line_index(self.message.start());
        let mut start_index = content_start_index.saturating_sub(2);

        // Trim leading empty lines.
        while start_index < content_start_index {
            if !source_code.line_text(start_index).trim().is_empty() {
                break;
            }
            start_index = start_index.saturating_add(1);
        }

        let content_end_index = source_code.line_index(self.message.end());
        let mut end_index = content_end_index
            .saturating_add(2)
            .min(OneIndexed::from_zero_indexed(source_code.line_count()));

        // Trim trailing empty lines.
        while end_index > content_end_index {
            if !source_code.line_text(end_index).trim().is_empty() {
                break;
            }

            end_index = end_index.saturating_sub(1);
        }

        let start_offset = source_code.line_start(start_index);
        let end_offset = source_code.line_end(end_index);

        let source = source_code.slice(TextRange::new(start_offset, end_offset));
        let message_range = self.message.range() - start_offset;

        let start_byte = source[TextRange::up_to(message_range.start())]
            .chars()
            .map(|c| c.len_utf8())
            .sum();
        let end_byte = source[TextRange::up_to(message_range.end())]
            .chars()
            .map(|c| c.len_utf8())
            .sum();

        let mut code = self.message.code.bold().bright_red();

        // Disable colours for tests, if the user requests it via env var, or non-tty
        if cfg!(test) || !colored::control::SHOULD_COLORIZE.should_colorize() {
            code = code.clear();
        };

        let source_text = source.show_nonprinting();

        let snippet = Level::None.title("").snippet(
            Snippet::source(&source_text)
                .line_start(start_index.get())
                .annotation(Level::Error.span(start_byte..end_byte).label(&code)),
        );

        let snippet_with_footer = if let Some(s) = self.message.suggestion() {
            snippet.footer(Level::Help.title(s))
        } else {
            snippet
        };

        // Disable colours for tests, if the user requests it via env var, or non-tty
        let renderer = if !cfg!(test) && colored::control::SHOULD_COLORIZE.should_colorize() {
            Renderer::styled()
        } else {
            Renderer::plain()
        };
        let source_block = renderer.render(snippet_with_footer);
        writeln!(f, "{source_block}")
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::TextEmitter;
    use crate::settings::UnsafeFixes;

    #[test]
    fn default() {
        let mut emitter = TextEmitter::default().with_show_source(true);
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }

    #[test]
    fn fix_status() {
        let mut emitter = TextEmitter::default()
            .with_show_fix_status(true)
            .with_show_source(true);
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }

    #[test]
    fn fix_status_unsafe() {
        let mut emitter = TextEmitter::default()
            .with_show_fix_status(true)
            .with_show_source(true)
            .with_unsafe_fixes(UnsafeFixes::Enabled);
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
