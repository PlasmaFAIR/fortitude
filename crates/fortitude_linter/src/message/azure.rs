// Adapted from from ruff
// Copyright 2022 Charles Marsh
// SPDX-License-Identifier: MIT

use std::io::Write;

use crate::message::Emitter;

use super::DiagnosticMessage;

/// Generate error logging commands for Azure Pipelines format.
/// See [documentation](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logissue-log-an-error-or-warning)
#[derive(Default)]
pub struct AzureEmitter;

impl Emitter for AzureEmitter {
    fn emit(
        &mut self,
        writer: &mut dyn Write,
        messages: &[DiagnosticMessage],
    ) -> anyhow::Result<()> {
        for message in messages {
            let location = message.compute_start_location();

            writeln!(
                writer,
                "##vso[task.logissue type=error\
                        ;sourcepath={filename};linenumber={line};columnnumber={col};{code}]{body}",
                filename = message.filename(),
                line = location.row,
                col = location.column,
                code = message
                    .rule()
                    .map_or_else(String::new, |rule| format!("code={};", rule.noqa_code())),
                body = message.body(),
            )?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::AzureEmitter;

    #[test]
    fn output() {
        let mut emitter = AzureEmitter;
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
