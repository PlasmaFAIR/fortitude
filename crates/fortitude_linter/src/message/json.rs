// Adapted from from ruff
// Copyright 2022 Charles Marsh
// SPDX-License-Identifier: MIT

use std::io::Write;

use serde::ser::SerializeSeq;
use serde::{Serialize, Serializer};
use serde_json::{json, Value};

use ruff_diagnostics::Edit;
use ruff_source_file::SourceCode;
use ruff_text_size::Ranged;

use crate::message::{DiagnosticMessage, Emitter};

#[derive(Default)]
pub struct JsonEmitter;

impl Emitter for JsonEmitter {
    fn emit(
        &mut self,
        writer: &mut dyn Write,
        messages: &[DiagnosticMessage],
    ) -> anyhow::Result<()> {
        serde_json::to_writer_pretty(writer, &ExpandedMessages { messages })?;

        Ok(())
    }
}

struct ExpandedMessages<'a> {
    messages: &'a [DiagnosticMessage],
}

impl Serialize for ExpandedMessages<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_seq(Some(self.messages.len()))?;

        for message in self.messages {
            let value = message_to_json_value(message);
            s.serialize_element(&value)?;
        }

        s.end()
    }
}

pub(crate) fn message_to_json_value(message: &DiagnosticMessage) -> Value {
    let source_code = message.source_file().to_source_code();

    let fix = message.fix().map(|fix| {
        json!({
            "applicability": fix.applicability(),
            "message": message.suggestion(),
            "edits": &ExpandedEdits { edits: fix.edits(), source_code: &source_code },
        })
    });

    let start_location = source_code.source_location(message.start());
    let end_location = source_code.source_location(message.end());

    json!({
        "code": message.rule().map(|rule| rule.noqa_code().to_string()),
        "message": message.body(),
        "fix": fix,
        "location": start_location,
        "end_location": end_location,
        "filename": message.filename(),
    })
}

struct ExpandedEdits<'a> {
    edits: &'a [Edit],
    source_code: &'a SourceCode<'a, 'a>,
}

impl Serialize for ExpandedEdits<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_seq(Some(self.edits.len()))?;

        for edit in self.edits {
            let location = self.source_code.source_location(edit.start());
            let end_location = self.source_code.source_location(edit.end());

            let value = json!({
                "content": edit.content().unwrap_or_default(),
                "location": location,
                "end_location": end_location
            });

            s.serialize_element(&value)?;
        }

        s.end()
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::JsonEmitter;

    #[test]
    fn output() {
        let mut emitter = JsonEmitter;
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
