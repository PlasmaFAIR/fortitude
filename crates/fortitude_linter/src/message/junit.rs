// Adapted from from ruff
// Copyright 2022 Charles Marsh
// SPDX-License-Identifier: MIT

use std::io::Write;
use std::path::Path;

use quick_junit::{NonSuccessKind, Report, TestCase, TestCaseStatus, TestSuite, XmlString};

use crate::message::{group_messages_by_filename, Emitter, MessageWithLocation};

use super::DiagnosticMessage;

#[derive(Default)]
pub struct JunitEmitter;

impl Emitter for JunitEmitter {
    fn emit(
        &mut self,
        writer: &mut dyn Write,
        messages: &[DiagnosticMessage],
    ) -> anyhow::Result<()> {
        let mut report = Report::new("fortitude");

        if messages.is_empty() {
            let mut test_suite = TestSuite::new("fortitude");
            test_suite.extra.insert(
                XmlString::new("package"),
                XmlString::new("org.plasmafair.fortitude"),
            );
            let mut case = TestCase::new("No errors found", TestCaseStatus::success());
            case.set_classname("fortitude");
            test_suite.add_test_case(case);
            report.add_test_suite(test_suite);
        } else {
            for (filename, messages) in group_messages_by_filename(messages) {
                let mut test_suite = TestSuite::new(filename);
                test_suite.extra.insert(
                    XmlString::new("package"),
                    XmlString::new("org.plasmafair.fortitude"),
                );

                for message in messages {
                    let MessageWithLocation {
                        message,
                        start_location,
                    } = message;
                    let mut status = TestCaseStatus::non_success(NonSuccessKind::Failure);
                    status.set_message(message.body());
                    let location = start_location;

                    status.set_description(format!(
                        "line {row}, col {col}, {body}",
                        row = location.row,
                        col = location.column,
                        body = message.body()
                    ));
                    let mut case = TestCase::new(
                        if let Some(rule) = message.rule() {
                            format!("org.plasmafair.fortitude.{}", rule.noqa_code())
                        } else {
                            "org.plasmafair.fortitude".to_string()
                        },
                        status,
                    );
                    let file_path = Path::new(filename);
                    let file_stem = file_path.file_stem().unwrap().to_str().unwrap();
                    let classname = file_path.parent().unwrap().join(file_stem);
                    case.set_classname(classname.to_str().unwrap());
                    case.extra.insert(
                        XmlString::new("line"),
                        XmlString::new(location.row.to_string()),
                    );
                    case.extra.insert(
                        XmlString::new("column"),
                        XmlString::new(location.column.to_string()),
                    );

                    test_suite.add_test_case(case);
                }
                report.add_test_suite(test_suite);
            }
        }

        report.serialize(writer)?;

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::JunitEmitter;

    #[test]
    fn output() {
        let mut emitter = JunitEmitter;
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
