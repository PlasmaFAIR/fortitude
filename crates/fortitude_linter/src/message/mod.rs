pub use azure::AzureEmitter;
pub use github::GithubEmitter;
pub use gitlab::GitlabEmitter;
pub use grouped::GroupedEmitter;
pub use json::JsonEmitter;
pub use json_lines::JsonLinesEmitter;
pub use junit::JunitEmitter;
pub use pylint::PylintEmitter;
pub use rdjson::RdjsonEmitter;
pub use sarif::SarifEmitter;
pub use text::TextEmitter;

mod azure;
mod diff;
mod github;
mod gitlab;
mod grouped;
mod json;
mod json_lines;
mod junit;
mod pylint;
mod rdjson;
mod sarif;
mod text;

use std::collections::BTreeMap;
use std::io::Write;
use std::ops::Deref;

use crate::diagnostic_message::DiagnosticMessage;
use ruff_source_file::SourceLocation;

/// Display format for a [`DiagnosticMessage`]s.
///
/// The emitter serializes a slice of [`DiagnosticMessage`]'s and writes them to a [`Write`].
pub trait Emitter {
    /// Serializes the `messages` and writes the output to `writer`.
    fn emit(
        &mut self,
        writer: &mut dyn Write,
        messages: &[DiagnosticMessage],
    ) -> anyhow::Result<()>;
}

struct MessageWithLocation<'a> {
    message: &'a DiagnosticMessage,
    start_location: SourceLocation,
}

impl Deref for MessageWithLocation<'_> {
    type Target = DiagnosticMessage;

    fn deref(&self) -> &Self::Target {
        self.message
    }
}

fn group_messages_by_filename(
    messages: &[DiagnosticMessage],
) -> BTreeMap<&str, Vec<MessageWithLocation<'_>>> {
    let mut grouped_messages = BTreeMap::default();
    for message in messages {
        grouped_messages
            .entry(message.filename())
            .or_insert_with(Vec::new)
            .push(MessageWithLocation {
                message,
                start_location: message.compute_start_location(),
            });
    }
    grouped_messages
}

#[cfg(test)]
mod tests {
    use ruff_diagnostics::{Diagnostic, DiagnosticKind, Edit, Fix};
    use ruff_source_file::SourceFileBuilder;
    use ruff_text_size::{TextRange, TextSize};

    use crate::message::{DiagnosticMessage, Emitter};

    pub(super) fn create_messages() -> Vec<DiagnosticMessage> {
        let test_contents = r#"module test
implicit none

contains
  subroutine foo
    implicit none
  end subroutine
end module
"#;

        let superfluous_implicit_none = Diagnostic::new(
            DiagnosticKind {
                name: "SuperfluousImplicitNone".to_string(),
                body: "'implicit none' set on the enclosing module".to_string(),
                suggestion: Some("Remove unnecessary 'implicit none'".to_string()),
            },
            TextRange::new(TextSize::from(57), TextSize::from(70)),
        )
        .with_fix(Fix::unsafe_edit(Edit::range_deletion(TextRange::new(
            TextSize::from(57),
            TextSize::from(70),
        ))));

        let unnamed_end_statement = Diagnostic::new(
            DiagnosticKind {
                name: "UnnamedEndStatement".to_string(),
                body: "end statement should read 'end subroutine foo'".to_string(),
                suggestion: None,
            },
            TextRange::new(TextSize::from(73), TextSize::from(87)),
        );

        let test_source = SourceFileBuilder::new("test.f90", test_contents).finish();

        let file_2 = r"integer*4 foo; end";

        let star_kind = Diagnostic::new(
            DiagnosticKind {
                name: "StarKind".to_string(),
                body: "integer*4 is non-standard, use integer(4)".to_string(),
                suggestion: None,
            },
            TextRange::new(TextSize::from(7), TextSize::from(8)),
        );

        let file_2_source = SourceFileBuilder::new("star_kind.f90", file_2).finish();

        vec![
            DiagnosticMessage::from_ruff(&test_source, superfluous_implicit_none),
            DiagnosticMessage::from_ruff(&test_source, unnamed_end_statement),
            DiagnosticMessage::from_ruff(&file_2_source, star_kind),
        ]
    }

    pub(super) fn capture_emitter_output(
        emitter: &mut dyn Emitter,
        messages: &[DiagnosticMessage],
    ) -> String {
        let mut output: Vec<u8> = Vec::new();
        emitter.emit(&mut output, messages).unwrap();

        String::from_utf8(output).expect("Output to be valid UTF-8")
    }
}
