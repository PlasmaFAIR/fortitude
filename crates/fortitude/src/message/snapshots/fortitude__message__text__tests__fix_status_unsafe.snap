---
source: fortitude/src/message/text.rs
expression: content
snapshot_kind: text
---
test.f90:6:5: S201 [*] 'implicit none' set on the enclosing module
  |
4 | contains
5 |   subroutine foo
6 |     implicit none
  |     ^^^^^^^^^^^^^ S201
7 |   end subroutine
8 | end module
  |
  = help: Remove unnecessary 'implicit none'

test.f90:7:3: S061 end statement should read 'end subroutine foo'
  |
5 |   subroutine foo
6 |     implicit none
7 |   end subroutine
  |   ^^^^^^^^^^^^^^ S061
8 | end module
  |

star_kind.f90:1:8: PORT021 integer*4 is non-standard, use integer(4)
  |
1 | integer*4 foo; end
  |        ^ PORT021
  |
